src/devices/infactory.c

```diff
@@ -1,84 +1,127 @@
+- b: battery low // flag to indicate low battery voltage
+- h: Humidity // BCD-encoded, each nibble is one digit, 'A0' means 100%rH
+- t: Temperature // in °F as binary number with one decimal place + 90 °F offset
 - n: Channel // Channel number 1 - 3
 
 Usage:
 
     # rtl_433 -f 434052000 -R 91 -F json:log.json
+    # rtl_433 -R 91 -F json:log.json
+    # rtl_433 -C si
 
 Payload looks like this:
 
-    [00] { 4} 00             : 0000
-    [01] {40} 0f 30 5c e7 61 : 00001111 00110000 01011100 11100111 01100001
+    [00] {40} 0f 30 5c e7 61 : 00001111 00110000 01011100 11100111 01100001
 
-First row is actually the preamble part. This is added to make the signal more unique.
+(See below for more information about the signal timing.)
 */
 
 #include "decoder.h"
 
+static void infactory_raw_msg(char *msg, uint8_t *b) {
+    int i;
+    char c;
+    char *p = msg;
+    for (i = 0; i < 10; i++) {
+        c = ((b[i >> 1]) >> (4 * (1 - (i & 1)))) & 0x0F;
+        c = (c > 9) ? c + 55 : c + 48;
+        *p++ = c;
+    }
+    *p = 0;
+}
+
+static int infactory_crc_check(uint8_t *b) {
+    int i;
+    uint8_t msg_crc, crc;
+    uint8_t msg[5];
+    memcpy(msg, b, 5);
+    msg_crc = msg[1] >> 4;
+    // for CRC computation, channel bits are at the CRC position(!)
+    msg[1] = (msg[1] & 0x0F) | (msg[4] & 0x0F) << 4;
+    crc = msg[0] >> 4;
+    for (i = 4; i < 36; i++) {
+        crc <<= 1;
+        if (msg[i>>3] & (0x80 >> (i&7))) {
+            crc |= 1; 
+        }
+        if (crc & 0x10) {
+            crc ^= 0x13; // Koopmann 0x9, CCITT-4; FP-4; ITU-T G.704
+        }
+    }
+
+    return (crc == msg_crc);
+}
+
 static int infactory_callback(r_device *decoder, bitbuffer_t *bitbuffer)
 {
     bitrow_t *bb;
     data_t *data;
     uint8_t *b;
-    int id, humidity, temp_raw, channel;
+    int id, battery, temp_raw, humidity, channel;
     float temp_f;
+    char *mic = "ERROR";
+    char raw_msg[11] = ".........."; // 10 hex chars plus 0
 
-    if (bitbuffer->bits_per_row[0] != 4) {
-        return DECODE_ABORT_LENGTH;
-    }
-    if (bitbuffer->bits_per_row[1] != 40) {
+    if (bitbuffer->bits_per_row[0] != 40) {
         return DECODE_ABORT_LENGTH;
     }
     bb = bitbuffer->bb;
 
-    /* Check that 4 bits of preamble is 0 */
+    /* Check that the last 4 bits of message are not 0 (channel number 1 - 3) */
     b = bb[0];
-    if (b[0]&0x0F)
-        return DECODE_ABORT_EARLY;
-
-    /* Check that last 4 bits of message is 0x1 */
-    b = bb[1];
     if (!(b[4]&0x0F))
         return DECODE_ABORT_EARLY;
 
     id       = b[0];
-    humidity = (b[3] & 0x0F) * 10 + (b[4] >> 4); // BCD
+    battery  = (b[1] >> 2) & 1;
     temp_raw = (b[2] << 4) | (b[3] >> 4);
-    temp_f   = (float)temp_raw * 0.1 - 90;
+    humidity = (b[3] & 0x0F) * 10 + (b[4] >> 4); // BCD, 'A0'=100%rH
     channel  = b[4] & 0x03;
 
+    temp_f   = (float)temp_raw * 0.1 - 90;
+
+    mic = infactory_crc_check(b) ? "CRC" : "CRC_ERROR"; // we could also return DECODE_FAIL_MIC;
```

Comment:
Correct we only return payloads with crc ok. And we should check the crc before we decode the parameters. Fail as quick as possible.

Reply:
OK. Will change.